### PR TITLE
Add load balancer in front of apiserver in HA master setup

### PR DESCRIPTION
**Pull request by fgrzadkowski**
The first commit is just https://github.com/kubernetes/kubernetes/pull/29201 and has been already LGTMed.

Second commit has some small fixes:
1. Precompute REGION variable in config
2. Add timeout for waiting for loadbalancer
3. Fix kube-down so that it doesn't delete some resources if there are still masters/nodes in other zones

Second commit also fixes bug in https://github.com/kubernetes/kubernetes/pull/29201 where variable `REGION` was unset in `kube-down` when deleting master IP. The bug caused leaking of IP addresses. 

https://github.com/kubernetes/kubernetes/issues/21124

@davidopp @jszczepkowski @wojtek-t @mikedanese

### CODE REVIEW

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -807,7 +804,16 @@ function create-loadbalancer() {
...
-  until $(curl -k -m1 https://${KUBE_MASTER_IP} > /dev/null 2>&1); do echo -n .; done
+  local counter=0
+  until $(curl -k -m1 https://${KUBE_MASTER_IP} &> /dev/null); do
+    counter=$((counter+1))
+    echo -n .
+    if [[ ${counter} -ge 1800 ]]; then
+      echo -e "${color_red}TIMEOUT${color_norm}" >&2
+      echo -e "${color_red}Load balancer failed to initialize within ${counter} seconds.${color_norm}" >&2
+      exit 2
+    fi
```
Comment (roberthbailey) [inline cluster/gce/util.sh]: sleep here to not spam the k8s api endpoint?

**Comment by roberthbailey** [comment]
#29201 merged, so you can remove the first commit.

**Comment by fgrzadkowski** [comment]
#29201 got reverted in https://github.com/kubernetes/kubernetes/pull/29278. Here I'm adding it back (reverting the revert).

File: cluster/gce/util.sh (hunk 1 above)
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: During this loop those requests will timeout after 1s, so they will not run more often than once every second. Do you think we should make it less often?

**Comment by roberthbailey** [comment]
lgtm

**Comment by wojtek-t** [comment]
I have a strong feeling that this broke our tests again.

kubemark-500 just failed with:
```
11:33:58 + ./test/kubemark/run-e2e-tests.sh '--ginkgo.focus=\[Feature:Performance\]' --gather-resource-usage=true
11:33:59 Kubemark master name: kubemark-500-kubemark-master
11:33:59 Project: k8s-jenkins-blocking-kubemark
11:33:59 Zone: us-central1-f
11:33:59 ERROR: (gcloud.compute.addresses.describe) Could not fetch resource:
11:33:59  - The resource 'projects/k8s-jenkins-blocking-kubemark/regions/us-central1/addresses/kubemark-500-kubemark-master-ip' was not found
11:33:59 
11:33:59 + dump_cluster_logs
```
http://kubekins.dls.corp.google.com/view/Scalability/job/kubernetes-kubemark-500-gce/4096/console

Unless, this isn't some strange flake, we may need to revert it again...

@k8s-oncall @fgrzadkowski @roberthbailey @lavalamp ^^

**Comment by wojtek-t** [comment]
And this was the first one with this PR in.

**Comment by apelisse** [comment]
Yet it finishes with success?

**Comment by wojtek-t** [comment]
This is super strange - it is green, but we didn't even started test - I've filed:  https://github.com/kubernetes/kubernetes/issues/29392

**Comment by fgrzadkowski** [comment]
I see the problem. In regular scripts we are using a static IP address for master and I depend on this in this PR. In kubemark we don't, which breaks this assumption. I'll send a fix for this.

**Pull request merged**